RFC-0031-ZenDNN-Integration.md

```diff
@@ -0,0 +1,94 @@
+
+# ZenDNN Integration
+
+
+## **Author**
+* @naveenthangudu
+
+
+## **Summary**
+This document proposes an approach for integrating ZenDNN library into PyTorch. This integration will enable inference optimizations for deep learning workloads on AMD CPUs.
+
+
+## **Highlights**
+* Default build behavior remains un-altered.
+* A pre-build step will generate ZenDNN integration code.
+* ZenDNN integration is optimized for inference workloads.
+* New opt-in build flag (environmental variable) `USE_ZENDNN` will be added.
+* No Submodules are added. When `USE_ZENDNN` is set to 1, the following repositories will be downloaded into third_party folder.
+  * [ZenDNN](https://github.com/amd/ZenDNN)
+  * [AOCL BLIS](https://github.com/amd/blis)
+* ZenDNN optimizations are added as JIT Passes on frozen graph under `optimize_for_inference` API.
+* Support will be added for fp32 and bf16 data types.
+
+
+## **Motivation**
+The main goal of this integration is to realize optimal inference on AMD CPUs for PyTorch. ZenDNN is a library which enables performance improvements on AMD CPU architectures. Please find the repo and more details [here](https://github.com/amd/ZenDNN). To highlight the potential uplift using ZenDNN vs. oneDNN on AMD CPUs, we have benchmarked our fully integrated ZenDNN version using PyTorch v1.12 on a Genoa CPU (4th Generation AMD EPYC™ Processors).
+
+
+### **Benchmarking Configuration**
+|||
+|-----------------------------|-------|
+| PyTorch version             | v1.12  |
+| CPU                         | 4th Generation AMD EPYC™ Genoa CPU |
+| Number of physical cores    | 96   |
+| NUMA Nodes per Socket(NPS)  | 1    |
+| Model data type             | FP32 |
+
+### **Latency Performance - batch_size=1**
+
+<img src="./RFC-0031-assets/Latency.svg" width=40%>
+
+### **Throughput Performance - batch_size=640**
+
+<img src="./RFC-0031-assets/Throughput.svg" width=40%>
+
+
+## **Proposed Implementation**
+
+### **Overview**
+This approach references the [hipify](https://github.com/ROCm-Developer-Tools/HIPIFY) approach used by the AMD ROCm team for integration of ROCm libraries into PyTorch. By leveraging shared APIs between oneDNN and ZenDNN, a pre-build step will generate ZenDNN integration code wherever possible. By adding an opt-in build flag, default build behavior is un-altered. Required repositories are only downloaded when the build flag is enabled.
```

Comment:
The problem this RFC is trying to solve is created in it's entirety by mechanically renaming oneDNN API symbols instead of maintaining compatibility. Whole renaming exercise puts burden on Pytorch developers without clear upside.

Reply:
Approach like hipify is suggested in this [thread](https://github.com/pytorch/pytorch/pull/76242) for reducing the amount of code needed for ZenDNN integration.

Reply:
This is not the top option preferred by Pytorch developers though. I understand why renaming was necessary with HIP. Unlike CUDA though oneDNN is [an open standard](https://www.oneapi.io/spec/) and has [open-source implementation](https://github.com/oneapi-src/oneDNN) under permissive license adopted by multiple hardware vendors.